src/include/migraphx/op/gather.hpp

```diff
@@ -61,35 +62,60 @@ struct gather
 
     shape normalize_compute_shape(std::vector<shape> inputs) const
     {
-        check_shapes{inputs, *this}.has(2);
-        auto lens = inputs[0].lens();
-        auto type = inputs[0].type();
-        lens.erase(lens.begin() + axis);
-        if(not inputs[1].scalar())
+        check_shapes{inputs, *this, true}.has(2);
+        shape data    = inputs[0];
+        shape indices = inputs[1];
+        auto type     = data.type();
+        // If index_dims is dynamic, convert the data to dynamic too.
+        if(indices.dynamic())
         {
-            auto ind_lens = inputs[1].lens();
-            lens.insert(lens.begin() + axis, ind_lens.begin(), ind_lens.end());
+            data = data.to_dynamic();
         }
-
-        // for scalar output
-        if(lens.empty())
+        if(data.dynamic())
         {
-            return {type};
+            std::vector<shape::dynamic_dimension> dims{data.dyn_dims()};
+            dims.erase(dims.begin() + axis);
+
+            if(not indices.scalar())
+            {
+                std::vector<shape::dynamic_dimension> index_dims;
+                index_dims = indices.to_dynamic().dyn_dims();
+                dims.insert(dims.begin() + axis, index_dims.begin(), index_dims.end());
```

Comment:
Do `dims.insert(dims.end(), ...)` this makes it clearer you are putting the elements at the end.

Reply:
That's not correct; dimensions can be inserted in the middle of the list.